src/potato_cluster.erl

```diff
@@ -0,0 +1,233 @@
+
+json_get(Key, Map) when is_atom(Key) ->
+    maps:get(atom_to_binary(Key, utf8), Map).
+
+json_get_str(Key, Map) when is_atom(Key) ->
+    binary_to_list(maps:get(atom_to_binary(Key, utf8), Map)).
+
+json_put(Key, Value, Map) when is_atom(Key) ->
+    maps:put(atom_to_binary(Key, utf8), Value, Map).
+
+start_one_pop_verifier(VerifierArr, JsonConf, MyIndex, UdpServerId, OnEventFn) ->
+
+    MyAddress = (array:get(MyIndex, VerifierArr))#verifier_public_info.network_data,
+
+    {_NetworkAddress, MyNodeNetId} = MyAddress,
+
+    NetSendFn = fun(DestAddressList, MsgId, Data) -> 
+			gen_server:cast({global, UdpServerId}, {send, DestAddressList, {MyAddress, MsgId, Data} })
+		end,
+
+    ConfPrivateKey = default,
+
+    VerifierOnEventFn = fun(Code, Data) -> OnEventFn(MyNodeNetId, Code, Data) end,
+
+    ConfigData = {MyIndex, NetSendFn, VerifierOnEventFn, ConfPrivateKey},
+
+    gen_server:start_link({global, MyNodeNetId}, pop_verifier, {json_map, JsonConf, ConfigData}, []),
+
+    gen_server:cast({global, UdpServerId}, {add_node, MyNodeNetId, {global, MyNodeNetId}}),
+
+    MyNodeNetId.
+
+start_single_server_cluster(VerifierArr, JsonConf, ServerAddress, UdpServerId, OnEventFn) ->
+    {_Ip, Port} = ServerAddress,
+
+    UdpOnEventFn = fun(Code, Data) -> OnEventFn(UdpServerId, Code, Data) end,
+
+    gen_server:start_link({global, UdpServerId}, potato_udp, {Port, UdpOnEventFn}, []),
+
+    VerList = lists:filter(
+		fun(VerData) ->
+			{VerAddress, _} = VerData#verifier_public_info.network_data,
+			VerAddress == ServerAddress
+		end,
+		array:to_list(VerifierArr)),
+
+    IndexList = lists:map(
+		  fun(VerData) ->
+			  VerData#verifier_public_info.index
+		  end,
+		  VerList),
+
+    IdList = lists:map(
+		fun(VerIndex) ->
+			start_one_pop_verifier(VerifierArr, JsonConf, VerIndex, UdpServerId, OnEventFn)
+		end, 
+		IndexList),
+    IdList.
+
+
+start_cluster_from_json(JsonConf_0, OnEventFn) ->
+
+    JsonConf = 
+	case json_get(genesis_block_timestamp_sec, JsonConf_0) of
+
+	    <<"now">> ->
+		Time = erlang:system_time(second),
+		json_put(genesis_block_timestamp_sec, Time, JsonConf_0);
+
+	    _ ->
+		JsonConf_0
+	end,
+
+    
+    
+    VerifierArr = pop_verifier:make_verifier_array_from_json(JsonConf),
+
+    VerAddressList = lists:map(
+		       fun(VerData) ->
+			       {VerAddress, _} = VerData#verifier_public_info.network_data,
+			       VerAddress
+		       end,
+		       array:to_list(VerifierArr)),
+
+    VerAddressUniqueList = sets:to_list(sets:from_list(VerAddressList)),
+
+    ProcessData = lists:map(
+		    fun(Address) ->
+			    UdpServerId = {udp_server, Address},
+			    IdList = start_single_server_cluster(VerifierArr, JsonConf, Address, UdpServerId, OnEventFn),
+			    {UdpServerId, IdList}
+		    end, 
+		    VerAddressUniqueList),
+
+    {UdpServerIdList, VerifierIdList} = lists:unzip(ProcessData),
+
+    {UdpServerIdList, lists:flatten(VerifierIdList)}.
+    
+stop_cluster({UdpServerIdList, VerifierIdList}) ->
```

Comment:
nice cleanup routine!!

Reply:
never used lol :scream:

Reply:
It's used in your unit test :o

Reply:
oh, it's the other one that is not used then

i never did a unit test for cluster with web_server thing

naming could have definitely been done better...